src/pykeen/nn/emb.py

```diff
@@ -846,6 +847,79 @@ def _sample(rs: torch.LongTensor, k: int) -> torch.LongTensor:
     return rs[torch.randperm(rs.shape[0])[:k]]
 
 
+def tokenize_anchors(
+    edge_index: np.ndarray,
+    anchors: np.ndarray,
+    num_tokens: int,
+    max_iter: int = 10,
+) -> torch.LongTensor:
+    """
+    Tokenize entities by representing them as a bag of anchor entities.
+
+    The entities are chosen by shortest path distance.
+
+    :param edge_index: shape: (2, m)
+        the edge index
+    :param anchors: shape: (num_anchors,)
+        the IDs of anchors entities
+    :param num_tokens:
+        the number of anchor IDs to select for each entity
+
+    :return: shape: (num_entities, num_tokens), -1 <= res < num_anchors
+        the selected anchor IDs for each entity
+    """
+    # TODO: use graph library, such as igraph, graph-tool, or networkit
+    # infer shape
+    num_entities = edge_index.max().item() + 1
+    # create adjacency matrix
+    adjacency = scipy.sparse.coo_matrix(
+        (
+            np.ones_like(edge_index[0], dtype=bool),
+            tuple(edge_index),
+        ),
+        shape=(num_entities, num_entities),
+    )
+    # symmetric + self-loops
+    adjacency = adjacency + adjacency.transpose() + scipy.sparse.eye(num_entities, dtype=bool, format="coo")
```

Comment:
I think we don't need to add `adjacency.transpose()` here since the arriving edge_index for NodePiece already contains inverse edges (hence symmetric)

Reply:
they get removed before (to be sure that relation context tokenization also works if we do not train the model with inverse relations)